File: python_scripts/mains/super_main_from_existing_ROM.py
# -*- coding: utf-8 -*-
"""
Created on Mon Mar 25 16:13:50 2019

@author: matheus.ladvig, valentin resseguier
.


"""
import numpy as np
import math
import time
from main_from_existing_ROM import main_from_existing_ROM
from super_main_from_existing_ROM_Simulation import super_main_from_existing_ROM_Simulation
#import matplotlib.pyplot as plt
#def super_main_from_existing_ROM(vect_nb_modes,type_data,v_threshold,vect_modal_dt,\
#    no_subampl_in_forecast,vect_reconstruction,vect_adv_corrected):
#    pass
#    
    
def switch_type_data(argument):
    switcher = {
        'incompact3D_noisy2D_40dt_subsampl_truncated': [[1e-05],[False]],
        'DNS100_inc3d_2D_2018_11_16_blocks_truncated': [[1e-06],[True]], # BEST
        'turb2D_blocks_truncated': [[1e-05],[False,True]],
        'incompact3d_wake_episode3_cut_truncated': [[1e-06],[False]],
        'incompact3d_wake_episode3_cut': [[1e-06],[False]],
        'LES_3D_tot_sub_sample_blurred': [[1e-03],[True]],
        'inc3D_Re3900_blocks': [[1e-03],[True]],
        'inc3D_Re3900_blocks_truncated': [[1e-03],[True]],
        'DNS300_inc3d_3D_2017_04_02_NOT_BLURRED_blocks_truncated':[[1e-04],[True]] # BEST
       
    }
    return switcher.get(argument,[[0.0005],[False]])


if __name__ == '__main__':

    vect_nb_modes = [2,4,6,8]
#    vect_nb_modes = [8,6,4,2]
    no_subampl_in_forecast = False 
    vect_reconstruction = [False] # for the super_main_from_existing_ROM
    vect_adv_corrected = [True]
    test_fct = 'b'
    svd_pchol = 2
    choice_n_subsample = 'htgen2'
#    choice_n_subsample = 'auto_shanon'
    stochastic_integration = 'Ito'
    estim_rmv_fv = True
#    eq_proj_div_free = 1
#    vect_eq_proj_div_free = 2
    vect_eq_proj_div_free = [2]
    EV = True
    thrDtCorrect = False
    noBugSubsampl = False
    
    #nb_mutation_steps = 0                # Number of mutation steps in particle filter 
    vect_nb_mutation_steps = [-1]                # Number of mutation steps in particle filter 
#    vect_nb_mutation_steps = [30,0]                # Number of mutation steps in particle filter 


#                           DATASET 
#    type_data = 'incompact3D_noisy2D_40dt_subsampl_truncated'  #dataset to debug
    
    type_data = 'DNS300_inc3d_3D_2017_04_02_NOT_BLURRED_blocks_truncated' # Reynolds 300
    SECONDS_OF_SIMU = 70. #70. #0.5                    # We have 331 seconds of real PIV data for reynolds=300 beacuse we have 4103 files. --> ( 4103*0.080833 = 331).....78 max in the case of fake_PIV

#    type_data = 'DNS100_inc3d_2D_2018_11_16_blocks_truncated'  # Reynolds 100
#    SECONDS_OF_SIMU = 100. #70. #0.5                   
    
#%%   
#Important parameters    
#    Threshold used in the estimation of the optimal subsampling time step
#     if modal-dt = true,
#          (mimic the use of a) disctinct subsampling time step for the
#           differentials equations of distincts chronos
    
    # Get threshold and modal_dt
    if choice_n_subsample == 'auto_shanon' :
        v_threshold,vect_modal_dt = switch_type_data(type_data)
    else:
         v_threshold = [float('nan')]
         vect_modal_dt = [False]
#    if not( choice_n_subsample == 'auto_shanon'):
#        vect_modal_dt = [False]
    
    nb_period_test = math.nan
    nb_modes_max = np.max(vect_nb_modes)
#    variances = []
    for eq_proj_div_free in vect_eq_proj_div_free:
        for nb_mutation_steps in vect_nb_mutation_steps:
            for modal_dt in vect_modal_dt:
                for threshold in v_threshold:
                    for adv_corrected in vect_adv_corrected:
                        for reconstruction in vect_reconstruction:
                            for k in vect_nb_modes:
                                n_particle = 100
                                main_from_existing_ROM(k,threshold,type_data,nb_period_test,\
                                                       no_subampl_in_forecast,reconstruction,\
                                                       adv_corrected,modal_dt,n_particle,\
                                                       test_fct,svd_pchol,\
                                                       stochastic_integration,\
                                                       estim_rmv_fv,eq_proj_div_free,\
                                                       thrDtCorrect,\
                                                       noBugSubsampl,\
                                                       choice_n_subsample,EV,\
                                                       nb_mutation_steps,
                                                       SECONDS_OF_SIMU)
    

    
    


          
            
            
            
            
#    super_main_from_existing_ROM_Simulation(vect_nb_modes,type_data,v_threshold,vect_modal_dt,\
#                                             no_subampl_in_forecast,vect_reconstruction,vect_adv_corrected)
#            
            
            
            
            
            
            
            
            
            
            
            
            
    
    
    
    

File: mains/super_main_from_existing_ROM.m
function super_main_from_existing_ROM(...
    vect_nb_modes,type_data,v_threshold,vect_modal_dt,...
    no_subampl_in_forecast,vect_reconstruction,vect_adv_corrected,...
    decor_by_subsampl,eq_proj_div_free, svd_pchol,noise_type)
% Launch a set of simulations with a several set of parameters
% Especially several number of modes
%

% % Number of periods reconstructed
% nb_period_test = 9;% for DNS 300
% % nb_period_test = 5;%for DNS 3900

if nargin == 0
    init;
    global correlated_model
%     global choice_n_subsample;
    global stochastic_integration;
    global estim_rmv_fv;
    
    %% Number of modes for the the ROM
%     vect_nb_modes = [ 16 8 6 4 2]
    vect_nb_modes = [4]
    no_subampl_in_forecast = false;
    vect_adv_corrected = [ true false ]
    vect_svd_pchol = 2
%     vect_svd_pchol = svd_pchol
    
    %% Type of data
    % Other datasets (do not use)
    % % type_data = 'LES_3D_tot_sub_sample_blurred';
    % % type_data = 'incompact3d_wake_episode3_cut_truncated';
    % % type_data = 'inc3D_Re3900_blocks_truncated';
    % % type_data = 'inc3D_Re3900_blocks';
    % % type_data = 'incompact3D_noisy2D_40dt_subsampl';
    % % type_data = 'incompact3D_noisy2D_40dt_subsampl_truncated';
    % % type_data = 'inc3D_Re300_40dt_blocks_truncated';
    % % type_data = 'turb2D_blocks_truncated'
    % % type_data = 'test2D_blocks_truncated'
    % % type_data = 'DNS300_inc3d_3D_2017_04_02_blocks_truncated'
    % % type_data = 'inc3D_HRLESlong_Re3900_blocks_truncated'
    % % type_data = 'small_test_in_blocks_truncated'
    % % type_data = 'test_1_block'
    % % type_data = 'inc3D_Re3900_blocks';
    % % type_data = 'turb2D_blocks_truncated'
    
    % These 3D data ( Re 300) gives good results
    % type_data = 'DNS300_inc3d_3D_2017_04_02_NOT_BLURRED_blocks_truncated'
    
    % These 2D data ( Re 100) gives good results
    type_data = 'DNS100_inc3d_2D_2018_11_16_blocks_truncated'
    
    % Small dataset for debuging
%     type_data = 'incompact3D_noisy2D_40dt_subsampl_truncated';
    % % type_data = 'incompact3d_wake_episode3_cut_truncated';
    
    % Data of the Re 100 OpenFOAM simulation
    % type_data = 'DNS100_OpenFOAM_2D_2020_blocks_truncated'
    % type_data = 'DNS100_OpenFOAM_2D_2020_blocks'
    % type_data = 'DNS100_OpenFOAM_2D_2020'
    
    if strcmp(type_data((end-9):end),'_truncated')
        vect_reconstruction = false;
    else
        vect_reconstruction = true;
    end
    
    %% Important parameters
    v_threshold=nan
    vect_modal_dt=false
    
    % (specific) time sub-sampling (forced time-decorrelation of unresolved chronos)
    decor_by_subsampl.bool=true;
    % Choice of subsampling time step based on chronos
    decor_by_subsampl.test_fct='b';
    % Way the subsampling is done (in which part of the code)
    % (can be  'bt_decor' or  'a_estim_decor')
    decor_by_subsampl.meth='bt_decor';
    % Meth to choose the time sub-sampling
    % ('auto_shanon'=maxim frequency of resolved chronos)
    % ('corr_time' = autocorrelation time estimation of the unresolved chronos)
    decor_by_subsampl.choice_n_subsample='htgen2'; % 'htgen' 'auto_shanon' 'lms'
    % decor_by_subsampl.choice_n_subsample = 'corr_time';
    decor_by_subsampl.bug_sampling = false ;
    
    % Stochastic integration path : 'Ito' or 'Str'
    stochastic_integration = 'Ito';
    
    % Definition of global variable to manage methods more easily
    choice_n_subsample = decor_by_subsampl.choice_n_subsample;
    
    % During the noise covariance estimation,
    % remove the finite-variation part of the chronos
    estim_rmv_fv = true;
    
    % Projection on the free-divergence-function space
    % 0 : no projection / 1 : projection of deterministic terms
    %  / 2 :projection of noise terms
    eq_proj_div_free = 2;
    
    correlated_model = false
else    
%     global choice_n_subsample;
    global stochastic_integration;
    global estim_rmv_fv;
    global correlated_model
    vect_svd_pchol = svd_pchol;
end

% % v_threshold=[1 10]/1000;
% % v_threshold=1e-3;
% if strcmp(type_data,'incompact3d_wake_episode3_cut_truncated')
%     % Number of periods reconstructed
%     nb_period_test = 9;% for DNS 300
% elseif strcmp(type_data,'LES_3D_tot_sub_sample_blurred')
%     % Number of periods reconstructed
%     nb_period_test = 5;%for DNS 3900
% else
    nb_period_test = nan;
% end

nb_modes_max = max(vect_nb_modes);

for svd_pchol=vect_svd_pchol
for modal_dt=vect_modal_dt
    for q=1:length(v_threshold)
        % parfor q=1:length(v_threshold)
        threshold=v_threshold(q);
        decor_by_subsampl_temp=decor_by_subsampl;
        decor_by_subsampl_temp.spectrum_threshold=threshold;
        for adv_corrected = vect_adv_corrected
            for reconstruction = vect_reconstruction
                % for k=vect_nb_modes
                for kk=1:length(vect_nb_modes)
                    k=vect_nb_modes(kk);
                    %     for k=nb_modes_min:2:nb_modes_max
                    main_from_existing_ROM(k,threshold,type_data,...
                        nb_period_test,...
                        no_subampl_in_forecast,reconstruction,...
                        adv_corrected,modal_dt,decor_by_subsampl_temp,...
                        svd_pchol,eq_proj_div_free,noise_type)
%                         adv_corrected,modal_dt,decor_by_subsampl.test_fct,...
%                         svd_pchol,eq_proj_div_free)
                    %         main_full_sto_vect_modal_dt_2nd_res(k,v_threshold(q))
                end
                %% Save plot
                %         folder_results = [ pwd '/resultats/current_results/'];
                %         current_pwd = pwd; cd ..
                %         folder_data = [ pwd '/data/' ];
                %         cd(current_pwd);
                %         % folder_results = ['/Users/Resseguier/Documents/MATLAB/POD/all/resultats/current_results/'];
                %         % folder_data = '/Users/Resseguier/Documents/MATLAB/POD/data/';
                %         % folder_results = ['/Users/vressegu/Documents/matlab/POD-NS_Stochastique/current_used/' ...
                %         %         'all/resultats/current_results/'];
                %         % %     param.folder_results =  [ pwd '/resultats/current_results/'];
                %
                %         threshold = num2str(q);
                %         iii = (threshold =='.');
                %         threshold(iii)='_';
                %
                %         str = ['print -depsc ' folder_results type_data '_sum_modes_n=' ...
                %             num2str(nb_modes_max) '_threshold_' threshold ...
                %             '_fullsto'];
                %         if vect_modal_dt
                %             str =[ str '_vect_modal_dt'];
                %         end
                %         if ~ adv_corrected
                %             str =[ str '_vect_modal_dt_NoAdvCorect'];
                %         end
                %         str =[ str '.eps'];
                %         eval(str);
                %         %     eval( ['print -depsc ' folder_results type_data '_sum_modes_n=' ...
                %         %         num2str(nb_modes_max) '_threshold_' num2str(v_threshold(q)) ...
                %         %         '_fullsto_vect_modal_dt.eps']);
                %         % %     eval( ['print -depsc ' folder_results 'sum_modes_n=' ...
                %         % %         num2str(nb_modes_max) '_threshold_' num2str(v_threshold(q)) '_fullsto_vect_modal_dt_test_basis.eps']);
                
            end
        end
    end
end
end

%% Plots
super_main_from_existing_ROM_Simulation(...
    vect_nb_modes,type_data,v_threshold,vect_modal_dt,...
    no_subampl_in_forecast,vect_reconstruction,vect_adv_corrected,...
    decor_by_subsampl,vect_svd_pchol,eq_proj_div_free,noise_type)
%     decor_by_subsampl.test_fct,vect_svd_pchol,eq_proj_div_free)

